### PR TITLE
Implement GET /swaps/:id (lightning)

### PR DESCRIPTION
**Pull request by tcharding**
Currently we do not return any data when a user hits the newly created `/swaps/:id` endpoints.

The schema was designed in https://github.com/comit-network/comit-js-sdk/pull/185

Return the correct data as defined in the js-sdk for the ethereum-ether/lightning endpoint.

Please note: Does not include tests as described in the issue.

Resolves: #2387

### CODE REVIEW

File: cnd/src/http_api/routes.rs [hunk 1]
```diff
@@ -60,7 +61,7 @@ pub async fn handle_get_halight_swap(
                 (alpha_ledger_state, beta_ledger_state, finalized_swap)
             }
             _ => {
-                let empty_swap = make_swap_entity(swap_id, vec![]);
+                let empty_swap = siren::Entity::default().with_class_member(http_api::PATH); // FIXME: Is this 'swaps' (http_api::PATH) correct?
```
Comment (thomaseizinger) [inline cnd/src/http_api/routes.rs:64]: This should not use the path, it has nothing to do with URLs: https://github.com/kevinswiber/siren#class

File: cnd/src/http_api/routes.rs [hunk 2]
```diff
@@ -99,25 +99,73 @@ pub async fn handle_get_halight_swap(
...
+    let swap = SwapResource {
+        status: state.status(),
+        role: (), // FIXME: use state.role() once serialization fore Role is done.
+    };
 
-    maybe_action_names
+    let entity = siren::Entity::default()
+        .with_class_member(http_api::PATH) // FIXME: Is this 'swaps' (http_api::PATH) correct?
+        .with_properties(swap)
+        // TODO: Add the entities?
```
Comment (thomaseizinger) [inline cnd/src/http_api/routes.rs:125]: Yes!

File: cnd/src/http_api/routes.rs [hunk 3]
```diff
@@ -99,25 +99,73 @@ pub async fn handle_get_halight_swap(
...
+    S: Status + MyRole,
+{
+    let swap = SwapResource {
+        status: state.status(),
+        role: (), // FIXME: use state.role() once serialization fore Role is done.
+    };
 
-    maybe_action_names
+    let entity = siren::Entity::default()
+        .with_class_member(http_api::PATH) // FIXME: Is this 'swaps' (http_api::PATH) correct?
```
Comment (thomaseizinger) [inline cnd/src/http_api/routes.rs:123]: No, it should just be a hardcoded string.

File: cnd/src/http_api/routes.rs [hunk 4]
```diff
@@ -99,25 +99,73 @@ pub async fn handle_get_halight_swap(
...
+        .with_properties(swap)
+        // TODO: Add the entities?
+        .map_err(|e| {
+            tracing::error!("failed to set properties of entity: {:?}", e);
+            HttpApiProblem::with_title_and_type_from_status(StatusCode::INTERNAL_SERVER_ERROR)
+        })?
+        // FIXME: Do we want a link to documentation of some sort?
```
Comment (thomaseizinger) [inline cnd/src/http_api/routes.rs:130]: Skip it for now I'd say.

File: cnd/src/http_api/routes.rs (hunk 2 above)
Comment (tcharding) [reply cnd/src/http_api/routes.rs]: I've pushed a bunch of code that starts with the properties.

File: cnd/src/http_api/routes.rs [hunk 5]
```diff
@@ -99,25 +100,108 @@ pub async fn handle_get_halight_swap(
...
+        quantity: state.quantity_wei(),
+    };
+    let alpha_params_sub = siren::SubEntity::from_entity(
+        siren::Entity::default()
+            .with_class_member("parameters")
+            .with_properties(alpha_params)
+            .map_err(|e| {
+                tracing::error!("failed to set properties of entity: {:?}", e);
+                HttpApiProblem::with_title_and_type_from_status(StatusCode::INTERNAL_SERVER_ERROR)
+            })?,
```
Comment (thomaseizinger) [inline cnd/src/http_api/routes.rs:145]:
```suggestion
            .context("failed to set alpha_params as properties for sub-entity")?,
```

File: cnd/src/http_api/routes.rs [hunk 6]
```diff
@@ -99,25 +100,108 @@ pub async fn handle_get_halight_swap(
...
+        &["beta"],
+    );
+    entity.push_sub_entity(beta_params_sub);
+
+    let alpha_tx = Transaction {
+        transactions: vec![],           // TODO: Add `StepTransactions`.
+        status: EscrowStatus::Deployed, // TODO: get this from somewhere.
```
Comment (thomaseizinger) [inline cnd/src/http_api/routes.rs:168]: That should be in the state.

File: cnd/src/http_api/routes.rs [hunk 7]
```diff
@@ -99,25 +100,108 @@ pub async fn handle_get_halight_swap(
...
+            .map_err(|e| {
+                tracing::error!("failed to set properties of entity: {:?}", e);
+                HttpApiProblem::with_title_and_type_from_status(StatusCode::INTERNAL_SERVER_ERROR)
+            })?,
+        &["beta"],
+    );
+    entity.push_sub_entity(beta_params_sub);
+
+    let alpha_tx = Transaction {
+        transactions: vec![],           // TODO: Add `StepTransactions`.
```
Comment (thomaseizinger) [inline cnd/src/http_api/routes.rs:167]: Should be in the state.

File: cnd/src/http_api/routes.rs [hunk 8]
```diff
@@ -148,6 +291,68 @@ pub struct BobHanEthereumHalightBitcoinState {
...
+impl QuantitySatoshi for BobHanEthereumHalightBitcoinState {
+    fn quantity_satoshi(&self) -> String {
+        self.finalized_swap.beta_asset.as_sat().to_string()
+    }
+}
+
+#[derive(Debug, Serialize)]
+struct HanEthereum {
+    pub protocol: String,
+    pub quantity: String, // In Wei.
```
Comment (thomaseizinger) [inline cnd/src/http_api/routes.rs:322]:
```suggestion
	 /// The quantity in wei.
    pub quantity: String,
```

File: cnd/src/http_api/routes.rs [hunk 9]
```diff
@@ -148,6 +291,68 @@ pub struct BobHanEthereumHalightBitcoinState {
...
+}
+
+#[derive(Debug, Serialize)]
+struct Transaction {
+    transactions: Vec<StepTransactions>,
+    status: EscrowStatus,
+}
+
+/// Key value store where values are transaction ids and keys are one of:
+/// "init", "deploy", "fund", "redeem", "refund"
```
Comment (thomaseizinger) [inline cnd/src/http_api/routes.rs:338]: Make it an enum then? :)

File: cnd/src/http_api/routes.rs (hunk 5 above)
Comment (tcharding) [reply cnd/src/http_api/routes.rs:145]: This suggestion doesn't build?

File: cnd/src/http_api/routes.rs (hunk 9 above)
Comment (tcharding) [reply cnd/src/http_api/routes.rs]: I was working under the assumption that changes to the API were out of scope for this ticket.  Is this assumption correct?  If not there are probably a few minor tweaks we can do to the schema defined in the js-sdk.  Would it be useful to write a `schema.json` file instead of coding to the js-sdk (or was that decided against already?)

File: cnd/src/http_api/routes.rs (hunk 6 above)
Comment (tcharding) [reply cnd/src/http_api/routes.rs]: Cool, I got it out using a `From` impl.

File: cnd/src/http_api/routes.rs (hunk 5 above)
Comment (thomaseizinger) [reply cnd/src/http_api/routes.rs:149]: You should be able to call `context` on anything that returns a Result where the `Error` implements `std::error::Error`.

Did I make an epic fail and return an error in the siren library that doesn't implement StdError?

File: cnd/src/http_api/routes.rs [hunk 10]
```diff
@@ -148,6 +322,184 @@ pub struct BobHanEthereumHalightBitcoinState {
...
+                transactions: HashMap::new(),
+                status: EscrowStatus::Funded,
+            },
+            halight::State::Settled(_) => Transaction {
+                transactions: HashMap::new(),
+                status: EscrowStatus::Redeemed,
+            },
+            halight::State::Cancelled(_) => Transaction {
+                transactions: HashMap::new(),
+                status: EscrowStatus::Refunded, // FIXME: Is this correct?
```
Comment (thomaseizinger) [inline cnd/src/http_api/routes.rs:497]: Yes I think that is correct.

File: cnd/src/http_api/routes.rs [hunk 11]
```diff
@@ -148,6 +322,184 @@ pub struct BobHanEthereumHalightBitcoinState {
...
+struct Transaction {
+    /// Keys are on of: "init", "deploy", "fund", "redeem", "refund".
+    /// Values are transactions.
+    transactions: HashMap<String, String>,
+    status: EscrowStatus,
+}
+
+#[derive(Debug, Clone, Copy, Serialize, PartialEq)]
+#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
+enum EscrowStatus {
```
Comment (thomaseizinger) [inline cnd/src/http_api/routes.rs:384]: Still not happy with these but I guess this is not the right place to discuss them.

Comment (tcharding) [reply cnd/src/http_api/routes.rs:384]: I also do not like this.  It seem to muddy together the two protocols (han/ethereum = halight) so I get the feeling that this is not going to be extensible. Pretty much all the code in `http_api/routes.rs` is horrible, its so entangled in the fact that its only this one protocol pair.

File: cnd/src/http_api/routes.rs (hunk 5 above)
Comment (tcharding) [reply cnd/src/http_api/routes.rs:149]:
```
rs:47:1
    |
47  |   pub enum EntityBuilderError {
    |   --------------------------- doesn't satisfy `_: anyhow::context::ext::StdError`
    |
    = note: the method `context` exists but the following trait bounds were not satisfied:
            `siren_types::types::EntityBuilderError: anyhow::context::ext::StdError`
            which is required by `std::result::Result<siren_types::types::Entity, siren_types::types::EntityBuilderError>: anyhow::Context<siren_types::types::Entity, siren_types::types::EntityBuilderError>`
    = help: items from traits can only be used if the trait is in scope
    = note: the following trait is implemented but not in scope; perhaps add a `use` for it:
            `use quick_error::ResultExt;`
```

Comment (tcharding) [reply cnd/src/http_api/routes.rs:149]: LOL, I just went to raise an issue on the siren-types repo to find it was ours - face palm.

Comment (tcharding) [reply cnd/src/http_api/routes.rs:149]: Why is siren-types under CoBloX organisation and not comit-network?

Comment (tcharding) [reply cnd/src/http_api/routes.rs:149]: Raised an issue on siret-types repo to track this.  https://github.com/coblox/rust-siren-types/issues/5

Comment (thomaseizinger) [reply cnd/src/http_api/routes.rs:149]: Because it doesn't really have anything to do with COMIT specifically, it is just a library :)

Comment (thomaseizinger) [reply cnd/src/http_api/routes.rs:149]: Epic fail Thomas! Why does this not implement `std::error::Error`!

File: api_tests/src/actors/actor.ts [hunk 12]
```diff
@@ -782,20 +782,14 @@ export class Actor {
             | "REFUNDED"
             | "INCORRECTLY_FUNDED"
     ) {
+        // This is hackery of the higest order.
```
Comment (thomaseizinger) [inline api_tests/src/actors/actor.ts:785]: Shouldn't exactly this PR allow to remove this hackery? We are returning the state now.

File: api_tests/src/actors/actor.ts [hunk 13]
```diff
@@ -807,6 +801,7 @@ export class Actor {
 
         do {
             swapEntity = await this.swap.fetchDetails();
+            console.log(JSON.stringify(swapEntity, null, 2));
```
Comment (thomaseizinger) [inline api_tests/src/actors/actor.ts:804]: This should not be part of this patch set :)

Comment (tcharding) [reply api_tests/src/actors/actor.ts:804]: Debugging code does not go into production, bad Tobin no biscuit.

File: api_tests/src/actors/actor.ts (hunk 12 above)
Comment (tcharding) [reply api_tests/src/actors/actor.ts:785]: This function is rcf003 specific but is called by code that uses the new protocols i.e., `bob.fund()` etc.  The test code needs detangling.

**Comment by tcharding** [comment]
> Since we are not adding any e2e test, we could add Rust tests for the serialization to at least have some idea of what it looks like :)

I feel we should have a JSON schema file, did not write it since we need to iterate on the API (in my opinion).

**Comment by tcharding** [comment]
Is this mergable now?

bors r+
